clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/CachedKafkaMetadataGetter.java

```diff
@@ -78,38 +80,64 @@ boolean containsTopic(TopicManager topicManager, String topicName) {
    * @param key cache key: Topic name or TopicPartition
    * @param metadataCache cache for this specific metadata
    * @param valueSupplier function to fetch metadata from Kafka
-   * @param <K> type of the cache key
    * @param <T> type of the metadata
    * @return the cache value or the fresh metadata from Kafka
    */
-  private <K, T> T fetchMetadata(K key, Map<K, ValueAndExpiryTime<T>> metadataCache, Supplier<T> valueSupplier) {
+  <T> T fetchMetadata(
+      KafkaMetadataCacheKey key,
+      Map<KafkaMetadataCacheKey, ValueAndExpiryTime<T>> metadataCache,
+      Supplier<T> valueSupplier) {
     final long now = System.nanoTime();
 
-    ValueAndExpiryTime<T> cachedValue = metadataCache.get(key);
+    final ValueAndExpiryTime<T> cachedValue = metadataCache.get(key);
     /**
      * The first entry of the pair is the expired time of this metadata; if the expired time is bigger than the current time,
      * reuse the cached value.
      */
     if (cachedValue != null && cachedValue.getExpiryTimeNs() > now) {
       return cachedValue.getValue();
     }
-    T newValue = valueSupplier.get();
+
     if (cachedValue == null) {
-      cachedValue = new ValueAndExpiryTime<>(newValue, now + ttlNs);
-      metadataCache.put(key, cachedValue);
-    } else {
-      // Reuse the existing object instead of creating a one to be more memory efficient.
-      cachedValue.setValueAndExpiryTimeMs(newValue, now + ttlNs);
+      T newValue = valueSupplier.get();
+      metadataCache.put(key, new ValueAndExpiryTime<>(newValue, now + ttlNs));
+      return newValue;
+    }
+
+    // TopicNotExistException is caught and thrown to upstream caller.
+    if (cachedValue.exception != null) {
+      TopicDoesNotExistException ex = cachedValue.exception;
+      cachedValue.exception = null;
+      throw ex;
+    }
+
+    // Update the value in the async fashion, for a given key and cache, we will only issue one async request at the
+    // same time.
+    if (cachedValue.valueUpdateInProgress.compareAndSet(false, true)) {
+      CompletableFuture.runAsync(() -> {
+        try {
+          T newValue = valueSupplier.get();
+          metadataCache.put(key, new ValueAndExpiryTime<>(newValue, System.nanoTime() + ttlNs));
+        } catch (TopicDoesNotExistException ex) {
+          cachedValue.exception = ex;
+          cachedValue.expiryTimeNs = System.nanoTime() + ttlNs;
```

Comment:
BTW, there is a behavior change by resetting expiry time.
Say, the metadata fetch hits an exception, the previous value preserves and expiry time gets reset.
The next request will throw an exception, and in the meantime, the exception gets reset.
The following requests will start returning the previous value since both exception is null and expiry time is not reached until the next refresh.

So this behavior is not quite consistent, maybe we could make it simpler.
1. Don't reset the exception when it is not null, and always throw the exception in the following requests until the next refresh.
2.  When hitting the expiry time, no matter whether exception is null or not, trigger an async refresh.

This will be a behavior change from the previous logic, but I think it might be fine and the behavior will be consistent.
Caching exception is another kind of back-off to avoid overwhelming the Kafka broker.